File: design_patterns/C++/Behavioral/chain_of_responsibility.cxx
#include <iostream>

class Account {
  protected:
	Account* successor = nullptr;
	float balance = 0.0;

  public:
	void setNext(Account* account) {
		this->successor = account;
	}

	void pay(float amountToPay) {
		if (this->canPay(amountToPay)) {
			std::cout << "Paid " << amountToPay << " using "
			          << typeid(*this).name() << "!" << std::endl;
		} else if (this->successor) {
			std::cout << "Cannot pay using " << typeid(*this).name()
			          << ". Proceeding..." << std::endl;
			this->successor->pay(amountToPay);
		} else {
			std::cout << "None of the accounts have enough balance" << std::endl;
		}
	}

	virtual bool canPay(float amountToPay) = 0;
};

class Bank: public Account {
  protected:
	float balance = 0.0;

  public:
	Bank(float balance) {
		this->balance = balance;
	}

	bool canPay(float amountToPay) {
		return this->balance >= amountToPay;
	}
};

class PayPal: public Account {
  protected:
	float balance = 0.0;

  public:
	PayPal(float balance) {
		this->balance = balance;
	}

	bool canPay(float amountToPay) {
		return this->balance >= amountToPay;
	}
};

class Bitcoin: public Account {
  protected:
	float balance = 0.0;

  public:
	Bitcoin(float balance) {
		this->balance = balance;
	}

	bool canPay(float amountToPay) {
		return this->balance >= amountToPay;
	}
};

int main() {
	// Let's prepare a chain like below
	//      $bank->$paypal->$bitcoin
	//
	// First priority bank
	//      If bank can't pay then paypal
	//      If paypal can't pay then bit coin
	auto bank = new Bank(100);
	auto paypal = new PayPal(200);
	auto bitcoin = new Bitcoin(300);
	bank->setNext(paypal);
	paypal->setNext(bitcoin);
	// Let's try to pay using the first priority i.e. bank
	bank->pay(259);
	// =======================================
	// Cannot pay using 4Bank. Proceeding...
	// Cannot pay using 6PayPal. Proceeding...
	// Paid 259 using 7Bitcoin!
	// =======================================
	return 0;
}

File: design_patterns/Python/Behavioral/chain_of_responsibility.py
class Account:

    def __init__(self):
        self.successor = None
        self.balance = 0

    def set_next(self, account):
        self.successor = account

    def pay(self, amount_to_pay):
        if self.can_pay(amount_to_pay):
            print('Paid {} using {}'.format(
                amount_to_pay, self.__class__.__name__))
        elif self.successor:
            print('Cannot pay using {}, Proceeding...'.format(
                self.__class__.__name__))
            self.successor.pay(amount_to_pay)
        else:
            raise Exception('None of the accounts have enough balance')

    def can_pay(self, amount):
        return self.balance >= amount


class Bank(Account):

    def __init__(self, balance):
        self.balance = balance


class PayPal(Account):

    def __init__(self, balance):
        self.balance = balance


class Bitcoin(Account):

    def __init__(self, balance):
        self.balance = balance


if __name__ == '__main__':

    # Let's prepare a chain like below
    #     bank->paypal->bitcoin
    # First priority bank
    # If bank can't pay then paypal
    # If paypal can't pay then bit coin
    bank = Bank(100)
    paypal = PayPal(200)
    bitcoin = Bitcoin(300)
    bank.set_next(paypal)
    paypal.set_next(bitcoin)

    bank.pay(259)
    # Cannot pay using Bank, Proceeding...
    # Cannot pay using PayPal, Proceeding...
    # Paid 259 using Bitcoin
